tests/tracer/test_uwsgi.py

```diff
@@ -0,0 +1,114 @@
+from collections import defaultdict
+import os
+import re
+import time
+
+import pytest
+
+from ddtrace.compat import httplib
+from ddtrace.vendor import six
+from tests import snapshot
+from tests.contrib.uwsgi import run_uwsgi
+
+
+uwsgi_app = os.path.join(os.path.dirname(__file__), "uwsgi-app.py")
+
+HTTP_PORT = 9091
+
+
+@pytest.fixture
+def uwsgi():
+    cmd = ["uwsgi", "--need-app", "--die-on-term", "--http-socket", ":{}".format(HTTP_PORT), "--wsgi-file", uwsgi_app]
+    yield run_uwsgi(cmd)
+
+
+@snapshot(async_mode=False)
+def test_uwsgi_single_request(uwsgi):
+    proc = uwsgi("--enable-threads", "--processes", "1", "--master")
+    # Give some time to child to actually startup
+    time.sleep(3)
+    try:
+        conn = httplib.HTTPConnection("localhost", HTTP_PORT)
+        conn.request("GET", "/")
+        resp = conn.getresponse()
+        assert resp.status == 200
+    finally:
+        proc.terminate()
+        proc.wait()
+
+
+@pytest.mark.skipif(six.PY2, reason="Does not work on Python2")
+@snapshot(async_mode=False)
+def test_uwsgi_multiple_requests(uwsgi):
+    proc = uwsgi("--enable-threads", "--processes", "2", "--master")
+    # Give some time to child to actually startup
+    time.sleep(3)
```

Comment:
ditto

Reply:
This same sleep is in the profiler uwsgi tests and seems like there's no way to get around having to wait before the tests make web requests to a server ready to accept connections.

Reply:
Yes, though the profiler tests do not try to send HTTP requests to uwsgi to see if it's up.
My thinking is that your approach here is better/safer and it should be the only one used, rather than having this sleep here.
I'd just replace the 10-tries approach with a `while True: … sleep(0.1)`.

Reply:
Took the advice and added a wait function to read the logs until I've detected the workers start. This would be a problem if it ends up waiting forever for the next line from the stdout, but CI would kill the job with its timeout in that case. What do you think?

Reply:
Awesome. I think you beat me to this since there's still sleep in the profiler tests 😢 

The CI will kill the job after 10 minutes if the test is broken and never stops. Since this should work in 100% of the case and it's not flaky and has no race condition, we should be good, right? 😉